### ISSUE TITLE
Add RSA key importing functionality for OpenSSL private keys

### ISSUE THREAD

**Issue by beamer159** [libtom/libtomcrypt#223]
TomCrypt currently supports importing OpenSSL RSA public keys. However, OpenSSL private keys cannot be imported currently. Adding this feature would expand this library's usability.

**Comment by karel-m** [comment]
Could you please post a sample key that you cannot import? Or even better an openssl  command used for "unimportable" key generation e.g. `openssl genrsa ...`?

**Comment by sjaeckel** [comment]
If you're talking about password-protected RSA private keys you're right. If you remove the password from the key you should be able to import it.

**Comment by karel-m** [comment]
If you generate an RSA key like this:
```
openssl genrsa -out test_rsa_private.pem 2048
openssl rsa -in test_rsa_private.pem -out test_rsa_private.der -outform der
openssl rsa -in test_rsa_private.pem -out test_rsa_public.der -outform der -pubout
```

libtomcrypt should be able to load `test_rsa_private.der` and `test_rsa_public.der` (if not then it's a bug).

**Comment by beamer159** [comment]
Here is the OpenSSL private key I am trying to open:

https://pastebin.com/nSZM1Gpt

Some things that I noticed:

For public keys, OpenSSL DER-encodes them differently from TomCrypt keys. A TomCrypt public key has the following structure:

    SEQUENCE
        INTEGER
        INTEGER

OpenSSL keys have a different structure:

    SEQUENCE
        SEQUENCE
            OBJECT IDENTIFIER
            NULL
        BIT STRING
            SEQUENCE
                INTEGER
                INTEGER

In rsa_import.c, I can find the code that converts the OpenSSL public key structure to the TomCrypt public key structure.

OpenSSL private keys are also encoded differently from TomCrypt private keys. Here's a TomCrypt private key structure:

    SEQUENCE
        INTEGER
        INTEGER
        INTEGER
        INTEGER
        INTEGER
        INTEGER
        INTEGER
        INTEGER
        INTEGER

And here's an OpenSSL private key structure:

    SEQUENCE
        INTEGER
        SEQUENCE
            OBJECT IDENTIFIER
            NULL
        BIT STRING
            SEQUENCE
                INTEGER
                INTEGER
                INTEGER
                INTEGER
                INTEGER
                INTEGER
                INTEGER
                INTEGER
                INTEGER

However, I don't see any code in rsa_import.c to convert this OpenSSL structure to a TomCrypt structure. It looks like rsa_import.c first tries to decode the key like an OpenSSL public key, then tries to decode the key like a TomCrypt public key, and finally tries to decode the like like a TomCrypt private key.

**Comment by karel-m** [comment]
Your keys has PKCS#8 format.

The function `rsa_import_pkcs8` from #157 is able to load it.

**Comment by beamer159** [comment]
That function is only provided in dev, correct? Not in the latest release, 1.17?

**Comment by karel-m** [comment]
Unfortunately it is not even in `develop` it's currently a pull requested.

Just take this file https://github.com/libtom/libtomcrypt/blob/fbcad861e47bd5e9fae047f1da0553adb352abdd/src/pk/rsa/rsa_import_pkcs8.c

**Comment by beamer159** [comment]
Thank you, I may have gotten that code working a bit, it's very helpful.

I have a question about some parts. The code has rsaoid which is an oid_st datatype. The code also uses corresponding functions pk_get_oid(...). This datatype and function are not in the latest current release, 1.17. It looks like this variable is only used to compare with something else. What would be the consequences of removing this check from my implementation?

**Comment by beamer159** [comment]
Is there any similar code to convert OpenSSL ECC private and public keys to be TomCrypt compliant?

**Comment by sjaeckel** [comment]
> Is there any similar code to convert OpenSSL ECC private and public keys to be TomCrypt compliant?

PR #65 has that, but it's also not merged yet as it's dependant on #187 which isn't merged yet

**Comment by sjaeckel** [comment]
I'm closing this one as it's a duplicate of #157
